File: src/backend/api/exps.py
from typing import List

from fastapi import APIRouter

from .. import exps
from .. import types
from .. import dates

router = APIRouter(prefix="/exps")


class WithTrainTestDates:
    train_start: types.PDTimestamp
    train_end: types.PDTimestamp
    test_start: types.PDTimestamp
    test_end: types.PDTimestamp


class ParsedExp(exps.Exp):
    type: str = "Exp"


class ParsedModelExp(exps.ModelExp, WithTrainTestDates):
    type: str = "ModelExp"


@router.get("/all")
async def getAllExps() -> List[ParsedExp | ParsedModelExp]:
    all_exps = []
    for exp in exps.all_exps:
        # the order if isinstance checks is important!
        if isinstance(exp, exps.ModelExp):
            exp_dates = dates.get_start_end_dates_for_exp(exp.exp_name)
            parsed_exp = ParsedModelExp(
                **{
                    **exp.model_dump(),
                    **{
                        "train_start": exp_dates.train_start,
                        "train_end": exp_dates.train_end,
                        "test_start": exp_dates.test_start,
                        "test_end": exp_dates.test_end,
                    },
                }
            )
            all_exps.append(parsed_exp)
        elif isinstance(exp, exps.Exp):
            all_exps.append(ParsedExp(**exp.model_dump()))
        else:
            raise ValueError(f"Unknown exp type: {exp}")
    return all_exps

File: src/frontend/src/api/exps.ts
import constants from "@/utils/constants";
import * as datetime from "@/utils/datetime";

interface ACType {
    mains: string;
    apps: string;
}

interface ResampleParams {
    rule: string;
    how: string;
    fill: string;
}

// enum for possible values of the parsed exp types: Exp, ModelExp
// see classes of exp in backend.exps
export enum ExpType {
    Exp = "Exp",
    ModelExp = "ModelExp",
}

// backend.api.exps.*
export interface ParsedExp {
    // backend.api.exps.ParsedExp
    type: ExpType;

    // backend.exps.Exp
    dataset: string;
    house: number;
    app: string;
    exp_name: string;
    app_norm_params: Record<string, number>;
    mains_norm_params: Record<string, number>;
    selected_ac_type: ACType;
    resample_params: ResampleParams;
    on_power_threshold: number;
    description: string;
}

export interface ParsedModelExp extends ParsedExp {
    // backend.exps.ModelExp
    selected_model_weights: string;
    sequence_length: number;
    selected_train_percent: number;
    batch_size: number;
    num_epochs: number;
    model_name: string;
    model_class: string;
    ds_class: string;

    // backend.api.exps.ParsedModelExp
    // backend.api.exps.WithTrainTestDates
    train_start: datetime.SimpleDateTimeString;
    train_end: datetime.SimpleDateTimeString;
    test_start: datetime.SimpleDateTimeString;
    test_end: datetime.SimpleDateTimeString;
}

export async function getAll(): Promise<ParsedExp[]> {
    const resp = await fetch(`${constants.backendApiUrl}/exps/all`);
    const respObj = await resp.json();
    const result = [];
    for (let i = 0; i < respObj.length; i++) {
        const exp = respObj[i];
        if (exp.type === ExpType.ModelExp) {
            exp.train_start = datetime.isoToSimple(exp.train_start);
            exp.train_end = datetime.isoToSimple(exp.train_end);
            exp.test_start = datetime.isoToSimple(exp.test_start);
            exp.test_end = datetime.isoToSimple(exp.test_end);

            result.push(exp as ParsedModelExp);
        }
        else if (exp.type === ExpType.Exp) {
            result.push(exp as ParsedExp);
        }
        else {
            throw new Error(`Unknown exp type: ${exp.type}`);
        }
    }
    return respObj as ParsedExp[];
}
